rust/private/rustc.bzl

```diff
@@ -310,13 +317,50 @@ def collect_inputs(
             for additional_input in linker_input.additional_inputs
         ]
 
+    # Compute linkstamps.
+    linkstamp_outs = []
+
+    if (crate_info.type in ("bin", "cdylib") and
+        # Are linkstamps supported by the C++ toolchain?
+        cc_common.is_enabled(feature_configuration = feature_configuration, feature_name = "linkstamps") and
+        # Is Bazel recent enough to support Starlark linkstamps?
+        hasattr(cc_common, "register_linkstamp_compile_action") and
+        # The current rule doesn't define _grep_includes attribute; this
+        # attribute is required for compiling linkstamps.
+        hasattr(ctx.attr, "_grep_includes")):
+        for dep in ctx.attr.deps:
+            if CcInfo in dep and dep[CcInfo].linking_context:
+                linking_context = dep[CcInfo].linking_context
+                for linkstamp in linking_context.linkstamps().to_list():
+                    # The linkstamp output path is based on the binary crate
+                    # name and the input linkstamp path. This is to disambiguate
+                    # the linkstamp outputs produced by multiple binary crates
+                    # that depend on the same linkstamp. We use the same pattern
+                    # for the output name as the one used by native cc rules.
+                    out_name = "_objs/" + crate_info.output.basename + "/" + linkstamp.file().path[:-len(linkstamp.file().extension)] + "o"
+                    linkstamp_out = ctx.actions.declare_file(out_name)
+                    linkstamp_outs.append(linkstamp_out)
+                    cc_common.register_linkstamp_compile_action(
+                        actions = ctx.actions,
+                        cc_toolchain = cc_toolchain,
+                        feature_configuration = feature_configuration,
+                        grep_includes = ctx.file._grep_includes,
+                        source_file = linkstamp.file(),
+                        output_file = linkstamp_out,
+                        compilation_inputs = linkstamp.hdrs(),
```

Comment:
Oh I just realized one thing. We need to make sure this action is reexecuted in all the situations when we would re-link the binary. That can be achieved by adding dependencies on all linker inputs (including the toolchain inputs for Rust linking and Rust compilation), and all source files of the current target. AFAIK the only way to test this is in a shell test that invokes Bazel inside, and we don't have good support for that here in rules_rust (and the support we have internally is incredibly complex). So I'd be fine with manually veryfing that linkstamps are recompiled when:

* any transitive Rust dependency is modified
* any transitive C++ dependency is modified
* linkstamp is modified (this should be true already)
* Rust toolchain is modified
* source file of the binary is modified

Reply:
Other than that this PR LGTM, thank you!

Reply:
Actually I can be a little less dramatic and just say to make sure that linkstamp action has all the inputs that the rust compile action does (of course rust compile action will have compiled linkstmaps as input in addition).